### DISCUSSION TITLE
datastructure for LIFO

### DISCUSSION THREAD

**Discussion by wenhuizhang** [zephyrproject-rtos/zephyr#5968] [General] [answered]
Hi, to whom it may concern:

I don't understand why LIFO and FIFO are all initialized with a queue please?
It seems like operation changes when LIFO, it uses prepend, and when FIFO it uses append. 
This makes the logic correct, however the names are little confusing.

https://github.com/zephyrproject-rtos/zephyr/blob/2251ea04a8eeef76f3bc4bbe58ea4132a2f67fad/include/kernel.h#L1937

```
struct k_lifo {
	struct k_queue _queue;
};

#define _K_LIFO_INITIALIZER(obj) \
	{ \
	._queue = _K_QUEUE_INITIALIZER(obj._queue) \
	}

#define K_LIFO_INITIALIZER DEPRECATED_MACRO _K_LIFO_INITIALIZER

```

**Comment by nashif** [comment]
@Vudentz can you please look into this, AFAIK you made changes in this area.

**Comment by Vudentz** [comment, marked as answer]
Well they both are using a queue as storage for the items, in fact both k_fifo and k_lifo could be replaced by k_queue, but initially lifo and fifo did not share a common storage so there was a lot of code duplication.
